spec/controllers/supervisor/subjects_controller_spec.rb

```diff
@@ -0,0 +1,386 @@
+require "rails_helper"
+
+RSpec.describe Supervisor::SubjectsController, type: :controller do
+  let(:supervisor) { create(:user, :supervisor) }
+  let(:valid_params) { {name: "Ruby", max_score: 100, estimated_time_days: 5} }
+  let(:subject_record) { create(:subject, valid_params) }
+
+  before do
+    allow(controller).to receive(:current_user).and_return(supervisor)
+    allow(controller).to receive(:logged_in?).and_return(true)
+    allow(I18n).to receive(:t).and_call_original
+  end
+
+  describe "#index" do
+    subject(:action) { get :index }
+
+    it "responds with status ok" do
+      # Tạo data trước khi test
+      create_list(:subject, 3)
+      
+      action
+      expect(response).to have_http_status(:ok)
+      # Debug: kiểm tra xem controller có thực sự chạy không
+      expect(assigns(:subjects)).to be_present
+      expect(assigns(:pagy)).to be_present
+    end
+
+    context "with search params" do
+      let!(:subject1) { create(:subject, name: "Ruby on Rails") }
+      let!(:subject2) { create(:subject, name: "JavaScript") }
+
+      it "filters subjects by name" do
+        get :index, params: { search: "Ruby" }
+        expect(assigns(:subjects)).to include(subject1)
+        expect(assigns(:subjects)).not_to include(subject2)
+      end
+
+      it "handles blank search" do
+        get :index, params: { search: "" }
+        expect(assigns(:subjects)).to include(subject1, subject2)
+      end
+
+      it "handles nil search" do
+        get :index, params: { search: nil }
+        expect(assigns(:subjects)).to include(subject1, subject2)
+      end
+    end
+
+    context "with pagination" do
+      before do
+        create_list(:subject, 25)
+      end
+
+      it "paginates results" do
+        get :index
+        expect(assigns(:subjects).size).to be <= Settings.ui.items_per_page
+      end
+    end
+  end
+
+  describe "#show" do
+    context "when subject exists" do
+      let(:subject_with_tasks) { create(:subject) }
+      let!(:task1) { create(:task, taskable: subject_with_tasks) }
+      let!(:task2) { create(:task, taskable: subject_with_tasks) }
+      
+      subject(:action) { get :show, params: {id: subject_with_tasks.id} }
+
+      it { action; expect(response).to have_http_status(:ok) }
+      it { action; expect(assigns(:tasks)).to eq(subject_with_tasks.tasks.ordered_by_name) }
```

Comment:
Các it này e mô tả messages cho nó nữa nhé, cho tường minh code
Fix tương tự cho những chỗ khác